### ISSUE TITLE
Sentence CHRF silently accepts single reference as string

### ISSUE THREAD

**Issue by bricksdont** [mjpost/sacrebleu#121]
Hi Matt, all

While trying to debug an issue related to sentence-level CHRF, I noticed that the compat function `sentence_chrf` expects a list of references as input, but happily accepts a single string (a list of characters). But then only uses the first character as the reference:

```
>>> from sacrebleu import sentence_chrf
>>> sentence_chrf("hi there", "hi there")
chrF2 = 0.455
>>> sentence_chrf("hi there", "h")
chrF2 = 0.455
```

While the `sentence_score` function for BLEU intercepts strings instead of lists of strings:

https://github.com/mjpost/sacrebleu/blob/65747b6ef6846d49548fd5139d2a6d118c7a384e/sacrebleu/metrics/bleu.py#L227

the one for CHRF does not:

https://github.com/mjpost/sacrebleu/blob/65747b6ef6846d49548fd5139d2a6d118c7a384e/sacrebleu/metrics/chrf.py#L125

Could this perhaps be changed / standardized for all metrics?

**Comment by ozancaglayan** [comment]
Thanks for reporting this. The PR synchronizes the behaviors across metrics.

**Comment by bricksdont** [comment]
A similar issue, I think, is that corpus CHRF expects `List[Iterable[str]]` as input, and silently accepts `List[str]` which will fail in `zip_longest`, but the error message will be confusing.

https://github.com/mjpost/sacrebleu/blob/65747b6ef6846d49548fd5139d2a6d118c7a384e/sacrebleu/metrics/chrf.py#L136

**Comment by ozancaglayan** [comment]
I did not understand the latter issue. The signature of `corpus_*` is the same both in BLEU and chrF. I accept that having a compat API and also object-oriented API at the same time is kind of confusing, and the latter should be supported with more examples/docs and maybe unit tests to discover potential flaws & argument inconsistencies.

**Comment by bricksdont** [comment]
In this function, there is already code to "add some robustness to the input arguments":

https://github.com/mjpost/sacrebleu/blob/65747b6ef6846d49548fd5139d2a6d118c7a384e/sacrebleu/metrics/chrf.py#L146

The code intercepts strings as inputs to this function and makes them into lists. Could it also intercept / give an error message if the references argument is a simple list `List[str]` instead of the expected `List[List[str]]`?

If the input is `List[str]`, i.e. a single reference where the user forgot to wrap in an extra `[]`, then this function will zip the hypothesis tokens which each character of the reference tokens, and the error message in `zip_longest` will be confusing (the streams were of different lengths, or similar).

Is that perhaps something worth changing as well?

Let me know if this is more clear.

**Comment by bricksdont** [comment]
Suggestion:

    if isinstance(ref_streams[0], str):
            ref_streams = [ref_streams]

**Comment by ozancaglayan** [comment]
oh I see. As I am not the original author of those "robustness" parts, I wasn't able myself as well to clearly enumerate which scenarios are handled by these `isinstance` checks. I would actually suggest to remove all these stuff and select a single stable signature for `corpus_*` methods, and assume that the users obey to the documentation.

**Comment by bricksdont** [comment]
Agreed, I totally understand your reasoning.

**Comment by ozancaglayan** [comment]
You suggest the latter for all metrics, right?

**Comment by bricksdont** [comment]
In principle yes, but for backwards compatibility that would be a problem I think and would not so much break existing code, as still work but silently produce the wrong results.

**Comment by ozancaglayan** [comment]
not removing but, didn't you suggest adding the following as a 3rd check, to fix the issue?

```
if isinstance(ref_streams[0], str):
        ref_streams = [ref_streams]
```

**Comment by bricksdont** [comment]
Hm, I don't think I understand sorry :(

I thought you were saying that you would not want to add another of those checks, rather explain to users how to use the functions correctly?

(Yes, I did suggest this check for corpus CHRF and just checked - works the same way for all metrics, yes.)

**Comment by ozancaglayan** [comment]
sorry, as a quick fix, I wanted to incorporate your suggestion rather than the API fix that I've previously suggested. Could you also paste a small example to exploit the current issue so that we can add it to the tests?

**Comment by bricksdont** [comment]
Sure, sth like this?

```python
>>> import sacrebleu
>>> hyp = "Es wird mit Gesamtkosten von 18.4 Millionen Franken gerechnet."
>>> ref = "Es werden Gesamtkosten von 18.4 Millionen Franken veranschlagt."
# correct
>>> sacrebleu.corpus_chrf(hyp, [[ref]])
chrF2 = 0.678
# incorrect, but there is an isinstance check for this
>>> sacrebleu.corpus_chrf(hyp, ref)
chrF2 = 0.678
# no check for this situation
>>> sacrebleu.corpus_chrf(hyp, [ref])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/net/cephfs/scratch/mathmu/map-volatility/venvs/sockeye3-cpu/lib/python3.6/site-packages/sacrebleu/compat.py", line 103, in corpus_chrf
    return metric.corpus_score(hypotheses, references)
  File "/net/cephfs/scratch/mathmu/map-volatility/venvs/sockeye3-cpu/lib/python3.6/site-packages/sacrebleu/metrics/chrf.py", line 158, in corpus_score
    raise EOFError("Source and reference streams have different lengths!")
EOFError: Source and reference streams have different lengths!
```

**Comment by ozancaglayan** [comment]
This is also happening with BLEU. I suggest keeping the latter issue as it is for now, and coming up with a well-defined API rather than adding another isinstance fix which would add a 3rd possibility for the `ref_streams` argument's type.

**Comment by ozancaglayan** [comment]
By the way, 

```
>>> sentence_chrf("hi there", ["h"])
chrF2 = 0.455
```

This score is wrong as well. The official chrF script gives 7.57 instead of 45.5 for this. In `refactor2021` the numbers match probably because I changed how the chrF stats were computed.

**Comment by bricksdont** [comment]
I don't understand why this score is wrong. Does this mean CHRF was wrong in SacreBLEU this entire time?

**Comment by ozancaglayan** [comment]
Probably not at corpus level, because for that we had unit tests. But when I try the following for your example, I don't get 45.5

```
git checkout https://github.com/m-popovic/chrF.git
echo 'hi there' > h
echo 'h' > r
python chrf/chrf++.py -H h -R r -nw 0 -s
```

**Comment by martinpopel** [comment]
The current code does smoothing by [computing the effective order](https://github.com/mjpost/sacrebleu/blob/7bd5d8/sacrebleu/metrics/chrf.py#L85), which was introduced by @mjpost when [adding chrF](https://github.com/mjpost/sacrebleu/commit/8d273661cccb9f4eeed3705646e579c48a79e0a1) to sacreBLEU. I guess this was taken from the original chrF implementation, but I cannot confirm it (see below).

When now skimming through the original [chrF paper (2015)](https://www.aclweb.org/anthology/W15-3049.pdf) and the [chrF++ paper (2017)](https://www.statmt.org/wmt17/pdf/WMT70.pdf), I could not find anything about smoothing.

The current chrF++ implementation does not compute the effective order, instead it does smoothing by [the 1e-16 constant](https://github.com/m-popovic/chrF/blob/6d3c38/chrF%2B%2B.py#L98-L107).
The chrF++ code seems to simulate the original chrF metric by `-nw 0` (and chrF+ by `-nw 1`), but I am not sure if this is the only difference between chrF and chrF++.
Unfortunately, I was not able to find the original chrF implementation. I weakly remember @m-popovic uploaded the script on her old website (not on GitHub, nor any other versioning system), but I could not find that website anymore.

**Comment by ozancaglayan** [comment]
yes I didn't notice any kind of smoothing either in the chrF++. This explains why in my current branch where I synced things from chrF++, the numbers are the same.

The only difference between chrF and chrF++ is `-nw` being `2` in the latter.
I don't know what would be the correct way of proceeding here.

**Comment by martinpopel** [comment]
> I didn't notice any kind of smoothing either in the chrF++

But there is [the 1e-16 smoothing](https://github.com/m-popovic/chrF/blob/6d3c38/chrF%2B%2B.py#L98-L107) as I wrote above.
The same is used in [NLTK](https://github.com/nltk/nltk/blob/develop/nltk/translate/chrf_score.py#L113), something different in [Moses](https://github.com/moses-smt/mosesdecoder/blob/master/mert/CHRFScorer.cpp) (also default beta is 3 here instead of 2).

> The only difference between chrF and chrF++ is -nw being 2 in the latter.

According to the papers yes. But we cannot be sure before inspecting the original chrF implementation, which I could not find anymore.

> I don't know what would be the correct way of proceeding here.

Perhaps stay in sync with [chrF++ and chrF as implemented in the GitHub repo chrF](https://github.com/m-popovic/chrF).

**Comment by ozancaglayan** [comment]
We can have the effective order configurable as well similar to BLEU.

**Comment by ozancaglayan** [comment]
But then which behavior to keep for >= 2.0.0, that is the question :)

**Comment by ozancaglayan** [comment]
Since none of the other tools (i.e. chrF++, NLTK and Moses) applied effective order smoothing, I am now slightly leaning towards dropping it and adopting the same approach as chrF++. This'll change scores for sentence-level chrF with sacreBLEU >= 2.0.0 but I think it's important to be in sync with other implementations. (The new scores should still correlate perfectly with old ones I presume)

**Comment by bricksdont** [comment]
Yes, effective order is basically just another smoothing method. I asked Maja for her opinion on this, perhaps we get a reply before version 2.0.0 happens.

**Comment by ozancaglayan** [comment]
Returning back to the original issue, my plan for >=2.0.0 is to disallow all kinds of obscure input type manipulations and expect the user to obey to the type signatures.

```
In [12]: sentence_chrf("hi there", "hi there")
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-12-91fdfab4b6df> in <module>
----> 1 sentence_chrf("hi there", "hi there")

~/git/sacrebleu/sacrebleu/compat.py in sentence_chrf(hypothesis, references, char_order, word_order, beta, remove_whitespace)
    135         beta=beta,
    136         whitespace=not remove_whitespace)
--> 137     return metric.sentence_score(hypothesis, references)
    138 
    139 

~/git/sacrebleu/sacrebleu/metrics/chrf.py in sentence_score(self, hypothesis, references)
    264         :return: A `CHRFScore` object.
    265         """
--> 266         return super().sentence_score(hypothesis, references)
    267 
    268     def corpus_score(self, hypotheses: Sequence[str],

~/git/sacrebleu/sacrebleu/metrics/base.py in sentence_score(self, hypothesis, references)
    149         :return: A `Score` object.
    150         """
--> 151         self._check_sentence_score_args(hypothesis, references)
    152 
    153         stats = self._extract_corpus_statistics(

~/git/sacrebleu/sacrebleu/metrics/base.py in _check_sentence_score_args(self, hyp, refs)
    107 
    108         if err_msg:
--> 109             raise RuntimeError(f'{prefix}: {err_msg}')
    110 
    111     def _check_corpus_score_args(self, hyps: Sequence[str],

RuntimeError: CHRF: The argument `refs` should be a sequence of strings.
```

**Comment by RamoramaInteractive** [comment]
Is this error into Sacrebleu still present?

**Comment by martinpopel** [comment]
No, the error was fixed in 2.0.0, as indicated above by closing this issue.
When you try `sentence_chrf("hi there", "hi there") `, it now correctly fails with `RuntimeError: CHRF: The argument 'refs' should be a sequence of strings.`
You need to pass a list of references (`sentence_chrf("hi there", ["hi there"])`).
